src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs

```diff
@@ -1754,6 +1754,11 @@ public ErrorDiagnostic NestedRuntimePropertyAccessNotSupported(string? resourceS
                 TextSpan,
                 "BCP308",
                 $@"The decorator ""{decoratorName}"" may not be used on statements whose declared type is a reference to a user-defined type.");
+
+            public ErrorDiagnostic ValueCannotBeFlattened(TypeSymbol unflattenableValueType) => new(
+                TextSpan,
+                "BCP309",
+                $@"Values of type ""{unflattenableValueType.Name}"" cannot be flattened.");
```

Comment:
>flattened [](http://example.com/codeflow?start=78&length=9)

Can we add an explanation why the value can't be flattened. It'll help the user fix it faster by avoiding a hop to signature help or docs.